File: 2025/shinyoung/2025-08-03-jump-game.py
class Solution:
    def canJump(self, nums):
        # Greedy: start at end
        # Time: O(n)
        # Space: O(1)
        
        n=len(nums)
        target=n-1
        
        for i in range(n-1, -1, -1):
            max_jump=nums[i]
            
            if i+max_jump>=target:
                target=i
        return target == 0
    

solution = Solution()
print(solution.canJump([2, 3, 1, 1, 4]))
print(solution.canJump([3, 2, 1, 0, 4]))
            
        

File: 2025/JuyoungMoon/week9/2025-08-03-jump-game.cpp
// Author: Juyoung Moon

// KDD LeetCode Study Week 9: Greedy
// https://github.com/juyomo/leetcode-study

// LeetCode #55.
// https://leetcode.com/problems/jump-game/

class Solution {
public:
    bool canJump(vector<int>& nums) {
        int goal = nums.size() - 1;
        int start = goal - 1;

        while (start >= 0) {
            if (nums[start] >= goal - start) {
                goal = start;
                start = goal - 1;
            } else {
                start--;
            }
        }
        return goal == 0;
    }
};
